src/preferences/common/preference-service.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2017 Ericsson and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from 'inversify';
+import { Event, Emitter } from '../../application/common/event'
+import { PreferenceChangedEvent } from './preference-event'
+import { IPreferenceServer } from './preference-server'
+
+
+injectable()
+export interface IPreferenceService {
+    readonly onPreferenceChanged: Event<PreferenceChangedEvent>;
+
+    has(preferenceName: string): Promise<boolean>;
+
+    get<T>(preferenceName: string, defaultValue?: T): Promise<T>;
```

Comment:
There should be 2 overloads:
```ts
get<T>(preferenceName: string): Promise<T | undefined>
get<T>(preferenceName: string, defaultValue: T): Promise<T>
```
The first can return `undefined`, the second cannot. On the implementation side, you have only one method.

The same for all methods below.